pkg/lexer/lexer.go

```diff
@@ -0,0 +1,182 @@
+// Copyright © 2021 Rak Laptudirm <raklaptudirm@gmail.com>
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package lexer
+
+import (
+	"errors"
+	"unicode/utf8"
+
+	"github.com/raklaptudirm/mash/pkg/token"
+)
+
+// Various error values returned by lexer.consume.
+var (
+	ErrIllegalNUL = errors.New("illegal character NUL")
+	ErrIllegalBOM = errors.New("illegal byte order mark")
+	ErrIllegalEnc = errors.New("illegal utf-8 encoding")
+)
+
+// lexer represents a mash source string and related lexing information.
+type lexer struct {
+	src string // source string
+	ch  rune   // current character
+
+	insertSemi bool
+
+	Tokens chan token.Token // lexer token channel
+
+	err ErrorHandler // lexer errors handling function
+
+	offset   int // offset of the start of the token in the source
+	rdOffset int // offsen of the current rune in the source
+
+	start token.Position // position in the source of the start of the token
+	pos   token.Position // position in the source of the current rune
+
+	ErrCount int // number of errors encountered
+}
+
+const (
+	eof = -1     // end of file
+	bom = 0xFEFF // byte order mark
```

Comment:
Why make this complecated? Do we really need the byte order mark? What is the special use case you're considering?

Reply:
If I understand this correctly, this is only used to make sure that we are working with either with utf-8, utf-16 or utf-32 but since we are using rune already, there is no need for this.

Reply:
Byte order mark is an unicode non-character, which is used in utf-16 and utf-32 encode the byte order (endianness) of the provided encoding. It is unnecessary in utf-8, but allowed at the first character. Since I want to support all edge cases, I have added support for the byte order mark character.

Reply:
But that is precisely my point, I don't see an edge case when all you're doing is processing utf-8?

Reply:
The byte order mark is valid utf-8, and should be ignored by any implementation conforming to the unicode standard. Having it as the first rune is not an error, so the lexer should not report it as an illegal character, or process it as a command.

Reply:
I think we'll going around in circles - you do what you want in this case, I feel like this is redundant since we can also have commands that should be utf-8 valid characters (not sure what you mean) so this doesn't seem like an edge case.

Reply:
Should I resolve this @tjgurwara99?